GitHub actions: Make all Windows checks consistent

tautschnig opened the pull request:
With the exceptions of whether or not to create packages, run tests, or
do signing, make all three Windows configurations exactly the same.
Apart from cleanliness, the main benefit is the use of clcache for all
Windows build jobs, making them consistently faster.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
GitHub actions: Make all Windows checks consistent

With the exceptions of whether or not to create packages, run tests, or
do signing, make all three Windows configurations exactly the same.
Apart from cleanliness, the main benefit is the use of clcache for all
Windows build jobs, making them consistently faster:
check-vs-2019-build-and-test now completes after 36 minutes instead of
74 minutes.

tautschnig committed:
Use filter file to avoid Windows quoting problems

The filter expression passed to jq necessarily requires double quotes,
which the Windows shell transforms in mysterious ways. To avoid this
problem, place the filter expression in a file instead.

tautschnig committed:
Do not unnecessarily extend PATH

Adding elements to the PATH appears to result in failing lookups, when
the binary appears to be on the PATH already anyway.

tautschnig committed:
Use platform-specific path separator

Exclusion of known-broken tests from XML validation failed on Windows as
the path names use \ instead of /.

tautschnig commented:
@NlightNFotis @hannes-steffenhagen-diffblue @thomasspriggs This is finally ready for review.

hannes-steffenhagen-diffblue commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -239,21 +251,21 @@ jobs:
       - uses: actions/checkout@v2
         with:
-          submodules: recursive
+          submodules: true
```
I’d prefer if this stayed at recursive. It’s not needed right now becaue we don’t have submodules with submodules, but if we ever do then we won’t run into mysterious CI bugs then (well, not this particular one at any rate).

tautschnig replied:
I have added an additional commit to make all actions `recursive` as we previously had a rather arbitrary mix.

hannes-steffenhagen-diffblue commented on regression/validate-trace-xml-schema/check.py:
```diff
@@ -11,49 +11,47 @@
+    ['unsigned1', 'test.desc'],
+    ['reachability-slice-interproc3', 'test.desc'],
+    ['sync_lock_release-1', 'symbol_per_type.desc'],
+    # this test is marked smt-backend
```
No, I meant `broken-smt-backend`. I don’t think regular smt backend should be a problem. I actually don’t remember why the broken smt backend is a problem either because these shouldn’t pass the extra smt flag we do for the smt backend tests?

tautschnig replied:
My comment was misleading, I've fixed this now. What I meant to say (and the comment now does say): we cannot expect to run this test successfully as we don't use the SMT solver in the XML validation tests.

hannes-steffenhagen-diffblue reviewed (approved):
FWIW Github actions has a cache size limit of 5GB, one build of cbmc takes up ~120MB on my local machine. We need to keep an eye on the cache, for now it's probably fine but if gets thrashed all the time (which it easily may on days where we get a lot of PRs) caching is probably more of a pessimisation than optimisation.

tautschnig committed:
GitHub actions: make all submodule checkouts recursive

This isn't currently needed, but at least makes sure all actions are
consistent: we previously had some set to "true" while others were
marked "recursive."
